packages/remote-feature-flag-controller/src/utils/user-segmentation-utils.ts

```diff
@@ -0,0 +1,49 @@
+import type { Json } from '@metamask/utils';
+import { v4 as uuidV4 } from 'uuid';
+
+import type { FeatureFlagScopeValue } from '../remote-feature-flag-controller-types';
+
+/* eslint-disable no-bitwise */
+/**
+ * Generates a deterministic random number between 0 and 1 based on a metaMetricsId.
+ * This is useful for A/B testing and feature flag rollouts where we want
+ * consistent group assignment for the same user.
+ *
+ * @param metaMetricsId - The unique identifier used to generate the deterministic random number
+ * @returns A number between 0 and 1 that is deterministic for the given metaMetricsId
+ */
+export function generateDeterministicRandomNumber(
+  metaMetricsId: string,
+): number {
+  const hash = [...metaMetricsId].reduce((acc, char) => {
+    const chr = char.charCodeAt(0);
+    return ((acc << 5) - acc + chr) | 0;
+  }, 0);
+
+  return (hash >>> 0) / 0xffffffff;
+}
+
+/**
+ * Type guard to check if a value is a feature flag with scope.
+ * Used to validate feature flag objects that contain scope-based configurations.
+ *
+ * @param featureFlag - The value to check if it's a feature flag with scope
+ * @returns True if the value is a feature flag with scope, false otherwise
+ */
+export const isFeatureFlagWithScopeValue = (
+  featureFlag: Json,
+): featureFlag is FeatureFlagScopeValue => {
+  return (
+    typeof featureFlag === 'object' &&
+    featureFlag !== null &&
+    'scope' in featureFlag
+  );
+};
+
+/**
+ * Generates UUIDv4 as a fallback metaMetricsId
+ * @returns A UUIDv4 string
+ */
+export function generateFallbackMetaMetricsId(): string {
```

Comment:
This is non-deterministic, and therefore unsuitable for determining whether a feature flag should be enabled or not

Reply:
I guess the best way of handling this would be to accept a callback for generating the metametrics ID? Ideally we'd use a messenger action, but that would be tricky because of the separate metrics implementations, so a callback should work instead.

Reply:
Or as you've just pointed out in the extension PR @DDDDDanica , we can rely on `getMetaMetricsId` to always return a value (i.e. to generate if it's missing). But we'll still want to remove this fallback if we're expecting the callback to always return a value, and we'll want to update the type of the callback

Reply:
Discussed in slack thread, we will remove the callback and add this logic to client side if undefined 🙏🏻